File: scripts/build_docs_bundle.py
#!/usr/bin/env python3
"""Generate dashboard docs bundle JSON from markdown and binder metadata."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import yaml

REPO_ROOT = Path(__file__).resolve().parents[1]
BINDER_DIR = REPO_ROOT / "docs" / "_binder"
OUT_FILE = REPO_ROOT / "dashboard" / "src" / "generated" / "docs-bundle.json"


def load_yaml(path: Path) -> Any:
    """Load a YAML file and return the parsed object."""
    with path.open("r", encoding="utf-8") as fh:
        return yaml.safe_load(fh)


def main() -> int:
    """Build the dashboard docs bundle from binder metadata and markdown files."""
    nav = load_yaml(BINDER_DIR / "nav.yaml")
    classification = load_yaml(BINDER_DIR / "classification.yaml")
    migration = load_yaml(BINDER_DIR / "migration-map.yaml")

    contents: dict[str, str] = {}
    for section in nav.get("sections", []):
        for page in section.get("pages", []):
            page_path = page.get("path")
            if not page_path:
                continue
            abs_path = REPO_ROOT / page_path
            if not abs_path.exists():
                raise FileNotFoundError(f"Referenced docs file missing: {page_path}")
            contents[page_path] = abs_path.read_text(encoding="utf-8")

    bundle = {
        "version": 1,
        "nav": nav,
        "classification": classification,
        "migration": migration,
        "contents": contents,
    }

    OUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    OUT_FILE.write_text(json.dumps(bundle, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(f"Wrote docs bundle: {OUT_FILE.relative_to(REPO_ROOT)} ({len(contents)} pages)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: dashboard/scripts/build-docs-bundle.mjs
#!/usr/bin/env node

import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import yaml from "js-yaml";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const dashboardRoot = path.resolve(__dirname, "..");
const repoRoot = path.resolve(dashboardRoot, "..");
const binderDir = path.join(repoRoot, "docs", "_binder");
const outFile = path.join(dashboardRoot, "src", "generated", "docs-bundle.json");

async function exists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function readYaml(filePath) {
  const raw = await fs.readFile(filePath, "utf-8");
  const data = yaml.load(raw);
  if (!data || typeof data !== "object") {
    throw new Error(`Invalid YAML mapping: ${filePath}`);
  }
  return data;
}

async function main() {
  const navPath = path.join(binderDir, "nav.yaml");
  const classificationPath = path.join(binderDir, "classification.yaml");
  const migrationPath = path.join(binderDir, "migration-map.yaml");

  const binderAvailable =
    (await exists(navPath)) &&
    (await exists(classificationPath)) &&
    (await exists(migrationPath));

  if (!binderAvailable) {
    if (await exists(outFile)) {
      console.log(
        "Docs binder metadata is unavailable in this build context; using existing generated bundle.",
      );
      return;
    }
    throw new Error(
      "Cannot build docs bundle: docs/_binder files are missing and no generated bundle exists.",
    );
  }

  const nav = await readYaml(navPath);
  const classification = await readYaml(classificationPath);
  const migration = await readYaml(migrationPath);

  const contents = {};
  const sections = Array.isArray(nav.sections) ? nav.sections : [];

  for (const section of sections) {
    const pages = Array.isArray(section?.pages) ? section.pages : [];
    for (const page of pages) {
      const pagePath = page?.path;
      if (!pagePath || typeof pagePath !== "string") {
        continue;
      }
      const absPath = path.join(repoRoot, pagePath);
      if (!(await exists(absPath))) {
        throw new Error(`Referenced docs file missing: ${pagePath}`);
      }
      contents[pagePath] = await fs.readFile(absPath, "utf-8");
    }
  }

  const bundle = {
    version: 1,
    nav,
    classification,
    migration,
    contents,
  };

  await fs.mkdir(path.dirname(outFile), { recursive: true });
  await fs.writeFile(outFile, JSON.stringify(bundle, null, 2) + "\n", "utf-8");
  console.log(`Wrote docs bundle: ${path.relative(repoRoot, outFile)} (${Object.keys(contents).length} pages)`);
}

main().catch((error) => {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
});

